packages/editor/src/dataviews/actions/duplicate-post.tsx

```diff
@@ -0,0 +1,170 @@
+	RenderModal: ( { items, closeModal, onActionPerformed } ) => {
+		const [ item, setItem ] = useState( {
+			...items[ 0 ],
+			title: sprintf(
+				/* translators: %s: Existing template title */
+				__( '%s (Copy)' ),
+				getItemTitle( items[ 0 ] )
+			),
+		} );
+
+		const [ isCreatingPage, setIsCreatingPage ] = useState( false );
+
+		const { saveEntityRecord } = useDispatch( coreStore );
+		const { createSuccessNotice, createErrorNotice } =
+			useDispatch( noticesStore );
+
+		async function createPage( event: React.FormEvent ) {
+			event.preventDefault();
+
+			if ( isCreatingPage ) {
+				return;
+			}
+
+			const newItemOject = {
+				status: 'draft',
+				title: item.title,
+				slug: item.title || __( 'No title' ),
+				comment_status: item.comment_status,
+				content:
+					typeof item.content === 'string'
+						? item.content
+						: item.content.raw,
+				excerpt:
+					typeof item.excerpt === 'string'
+						? item.excerpt
+						: item.excerpt?.raw,
+				meta: item.meta,
+				parent: item.parent,
+				password: item.password,
+				template: item.template,
+				format: item.format,
+				featured_media: item.featured_media,
+				menu_order: item.menu_order,
+				ping_status: item.ping_status,
+			};
+			const assignablePropertiesPrefix = 'wp:action-assign-';
+			// Get all the properties that the current user is able to assign normally author, categories, tags,
+			// and custom taxonomies.
+			const assignableProperties = Object.keys( item?._links || {} )
+				.filter( ( property ) =>
+					property.startsWith( assignablePropertiesPrefix )
+				)
+				.map( ( property ) =>
+					property.slice( assignablePropertiesPrefix.length )
+				);
+			assignableProperties.forEach( ( property ) => {
+				if ( item.hasOwnProperty( property ) ) {
+					// @ts-ignore
+					newItemOject[ property ] = item[ property ];
+				}
+			} );
+			setIsCreatingPage( true );
+			try {
+				const newItem = await saveEntityRecord(
+					'postType',
+					item.type,
+					newItemOject,
+					{ throwOnError: true }
+				);
+
+				createSuccessNotice(
+					sprintf(
+						// translators: %s: Title of the created template e.g: "Category".
+						__( '"%s" successfully created.' ),
+						decodeEntities( newItem.title?.rendered || item.title )
+					),
+					{
+						id: 'duplicate-post-action',
+						type: 'snackbar',
+					}
+				);
+
+				if ( onActionPerformed ) {
+					onActionPerformed( [ newItem ] );
+				}
+			} catch ( error ) {
+				const typedError = error as CoreDataError;
+				const errorMessage =
+					typedError.message && typedError.code !== 'unknown_error'
+						? typedError.message
+						: __( 'An error occurred while duplicating the page.' );
+
+				createErrorNotice( errorMessage, {
+					type: 'snackbar',
+				} );
+			} finally {
+				setIsCreatingPage( false );
+				closeModal?.();
+			}
+		}
+
+		return (
+			<form onSubmit={ createPage }>
+				<VStack spacing={ 3 }>
+					<DataForm
+						data={ item }
+						fields={ fields }
```

Comment:
There's a typescript issue on this line, I'm not yet sure how to solve it.

Reply:
It seems like there are a couple of problems. A good place to start is by providing the type argument with the expected value. I don't think the expected `Item` type was being inferred correctly which makes the errors confusing:


```diff
-<DataForm
+<DataForm< typeof item >
 	data={ item }
```

It seems like there are problems with fields and onChange props.

Reply:
The way `item` is inferred here causes problems:

https://github.com/WordPress/gutenberg/blob/5fe12e9b79d0e3ef08b63af8e469eca7b06a8048/packages/editor/src/dataviews/actions/duplicate-post.tsx#L37-L44

That can be fixed by providing a type parameter to `useState`. I think this is what we want:

```diff
diff --git before/packages/editor/src/dataviews/actions/duplicate-post.tsx after/packages/editor/src/dataviews/actions/duplicate-post.tsx
index 5d0147691c..5b64f0ad6e 100644
--- before/packages/editor/src/dataviews/actions/duplicate-post.tsx
+++ after/packages/editor/src/dataviews/actions/duplicate-post.tsx
@@ -34,7 +34,7 @@ const duplicatePost: Action< BasePost > = {
 		return status !== 'trash';
 	},
 	RenderModal: ( { items, closeModal, onActionPerformed } ) => {
-		const [ item, setItem ] = useState( {
+		const [ item, setItem ] = useState< BasePost >( {
 			...items[ 0 ],
 			title: sprintf(
 				/* translators: %s: Existing template title */
```

Reply:
Two things that helped me fix this:
- when defining the state for `item`, declare the type explicitly with `useState<BasePost>`. Otherwise the type will be inferred from the object literal used to initialize the state, and that type is incomplete and a little chaotic.
- in `DataFormProps<Item>` type, change the type of `onChange` argument from `Record<string,any>` to `Item`. Because when `data` are `Item`s and `fields` are `Field<Item>`s, then the callback parameter is also `Item`, isn't it?

Reply:
no, the callback is not `Item`. Only the actual modified properties are passed to `onChange` because we want to know exactly which properties changed.

Reply:
It seems like there's currently a bug in `onChange` I'll fix it.

Reply:
Thanks all for your help, It's fixed now.

Reply:
> the callback is not `Item`. Only the actual modified properties are passed to `onChange`

In that case `Partial<Item>` should be the best type for the `onChange` parameter.

Reply:
I tried `Partial<Item>` but it doesn't work because of the `getValue` property of fields. In other words fields are not always `item[ id ]` but changes are.